core/src/kura.rs

```diff
@@ -536,9 +414,53 @@ pub struct BlockIndex {
     pub length: u64,
 }
 
-/// Operations available both on locked and unlocked store,
-/// i.e. only read operations.
-impl<L: Lock> BlockStore<L> {
+pub enum LockStatus {
+    Locked,
+    Unlocked,
+}
+
+impl BlockStore {
+    /// Create a new read-only block store in `path`.
+    pub fn new(path: &Path, already_locked: LockStatus) -> Self {
+        if matches!(already_locked, LockStatus::Unlocked) {
+            let path = path.join(LOCK_FILE_NAME);
+            if let Err(e) = fs::File::options()
+                .read(true)
+                .write(true)
+                .create_new(true)
+                .open(path.clone())
+            {
+                match e.kind() {
+                    _ => Err(Error::IO(e, path)),
+                    std::io::ErrorKind::AlreadyExists => Err(Error::Locked(path)),
+                    std::io::ErrorKind::NotFound => {
+                        match std::fs::create_dir_all(&path)
+                            .map_err(|e| Error::MkDir(e, path.clone()))
+                        {
+                            Err(e) => Err(e),
+                            Ok(_) => {
+                                if let Err(e) = fs::File::options()
+                                    .read(true)
+                                    .write(true)
+                                    .create_new(true)
+                                    .open(path.clone())
+                                {
+                                    Err(Error::IO(e, path))
+                                } else {
+                                    Ok(())
+                                }
+                            }
+                        }
+                    }
+                }
```

Comment:
```suggestion
                match e.kind() {
                    std::io::ErrorKind::AlreadyExists => Err(Error::Locked(path)),
                    std::io::ErrorKind::NotFound => std::fs::create_dir_all(&path)
                        .map_err(|e| Error::MkDir(e, path.clone()))
                        .and_then(|_| {
                            fs::File::options()
                                .read(true)
                                .write(true)
                                .create_new(true)
                                .open(&path)
                                .map_err(|e| Error::IO(e, path))
                        }),
                    _ => Err(Error::IO(e, path)),
                }
```
it seems matching have to be fixed, also you can simplify it using combinators and remove redundant clones.